pkg/dockerregistry/server/tagservice.go

```diff
@@ -0,0 +1,56 @@
+package server
+
+import (
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/digest"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+type tagService struct {
+	distribution.TagService
+
+	imageStream *imageapi.ImageStream
+}
+
+func (t tagService) Get(ctx context.Context, tag string) (distribution.Descriptor, error) {
+	context.GetLogger(ctx).Debug("(tagService).Get: starting")
+
+	te := imageapi.LatestTaggedImage(t.imageStream, tag)
+	if te == nil {
+		return distribution.Descriptor{}, distribution.ErrTagUnknown{Tag: tag}
+	}
+	dgst, err := digest.ParseDigest(te.Image)
+	if err != nil {
+		return distribution.Descriptor{}, err
+	}
+
+	return distribution.Descriptor{Digest: dgst}, nil
+}
+
+func (t tagService) All(ctx context.Context) ([]string, error) {
+	context.GetLogger(ctx).Debug("(tagService).All: starting")
+
+	tags := []string{}
+	for tag := range t.imageStream.Status.Tags {
+		tags = append(tags, tag)
+	}
+	return tags, nil
+}
+
+func (t tagService) Lookup(ctx context.Context, digest distribution.Descriptor) ([]string, error) {
+	context.GetLogger(ctx).Debug("(tagService).Lookup: starting")
+	return t.TagService.Lookup(ctx, digest)
+}
+
+// TODO(legion): Rewrite this method to store tag in etcd.
```

Comment:
> In the current version of docker-registry described path to tags and we do not have a wrappers for tag management in repositorymiddleware.go.

We already do with this file of yours. I'm affraid we don't have much choice here. `Untag` is easy using [ImageStreamTags.Delete](https://github.com/miminar/origin/blob/registry-quota-refactor/pkg/client/imagestreamtags.go#L48). `Lookup` should be easy as well - just going through the `IS.Spec`. `Tag` shouldn't be a problem either with [imageStreamTags.Update](https://github.com/miminar/origin/blob/registry-quota-refactor/pkg/client/imagestreamtags.go#L41).